manager/manager.go

```diff
@@ -1424,3 +1432,106 @@ func (m *Manager) newNetlinkConnection(ifindex int32, netns uint64) (*netlinkCac
 	m.netlinkCache[netlinkCacheKey{Ifindex: ifindex, Netns: netns}] = &cacheEntry
 	return &cacheEntry, nil
 }
+
+// cleanupKprobeEvents - Cleans up kprobe_events and uprobe_events by removing entries of known UIDs, that are not used
+// anymore.
+//
+// Previous instances of this manager might have been killed unexpectedly. When this happens,
+// kprobe_events is not cleaned up properly and can grow indefinitely until it reaches 65k
+// entries (see: https://elixir.bootlin.com/linux/latest/source/kernel/trace/trace_output.c#L696)
+// Once the limit is reached, the kernel refuses to load new probes and throws a "no such device"
+// error. To prevent this, start by cleaning up the kprobe_events entries of previous managers that
+// are not running anymore.
+func (m *Manager) cleanupTracefs() error {
+	// build the pattern to look for in kprobe_events and uprobe_events
+	var uids string
+	for _, p := range m.Probes {
+		if len(p.UID) == 0 {
+			continue
+		}
+
+		if len(uids) == 0 {
+			uids = p.UID
+		} else {
+			uids = uids + "|" + p.UID
+		}
+	}
+	pattern, err := regexp.Compile(fmt.Sprintf(`(p|r)[0-9]*:(kprobes|uprobes)/(.*(%s)_([0-9]*)) .*`, uids))
+	if err != nil {
+		return errors.Wrap(err, "pattern generation failed")
+	}
+
+	// clean up kprobe_events
+	pidMask := make(map[int]bool)
+	if err := cleanupKprobeEvents(pattern, pidMask); err != nil {
+		return err
+	}
+
+	// clean up uprobe_events
+	if err := cleanupUprobeEvents(pattern, pidMask); err != nil {
+		return err
+	}
+	return nil
+}
+
+func cleanupKprobeEvents(pattern *regexp.Regexp, pidMask map[int]bool) error {
+	kprobeEvents, err := ReadKprobeEvents()
+	if err != nil {
+		return errors.Wrap(err, "couldn't read kprobe_events")
+	}
+	for _, match := range pattern.FindAllStringSubmatch(kprobeEvents, -1) {
+		if len(match) < 6 {
+			continue
+		}
+
+		// check if the provided pid still exists
+		pid, err := strconv.Atoi(match[5])
+		if err != nil {
+			continue
+		}
+		if _, ok := pidMask[pid]; ok {
+			continue
+		}
+		_, err = process.NewProcess(int32(pid))
+		if err == nil {
+			// the process exists, continue
+			pidMask[pid] = true
+			continue
+		}
```

Comment:
How many kprobes do we have per process - is it worth caching whether this PID is alive in a map for reuse across probe checks?

Reply:
We have about 90 probes per process, and we've seen hosts with up to 60k processes. The goal was to try to avoid 5.4M /proc hits as much as possible 😓  ... I even think we should probably add a `time.Sleep(100*time.Millisecond)` to avoid a CPU spike on startup, what do you think ?

Reply:
What about storing the pid state `alive` `exited` so that there is no need to call `NewProcess` multiple times ?

Reply:
Adding `alive / exited` so that each is queried once 👍